contributors/design-proposals/cpu-affinity-numa-awareness.md

```diff
@@ -0,0 +1,216 @@
+The following `Taint` keys are defined to enable CPU pinning and NUMA awareness.
+
+#### CPUAffinity
+
+* Effect: `NoScheduleNoAdmitNoExecute`
+* Potential values:
+ * `dedicated`
+
+If `dedicated`, all pods that match this taint will require dedicated compute resources. Each
+pod bound to this node must request CPU in whole cores.  The CPU limit must equal the request.
+
+#### NUMACPUAffinity
+
+* Effect: `NoScheduleNoAdmitNoExecute`
+* Potential values:
+ * `strict`
+
+If `strict`, all pods that match this taint must request CPU (whole or fractional cores) that
+fit a single NUMA node `cpu` allocatable.
+
+#### NUMAMemoryPolicy
+
+* Effect: `NoScheduleNoAdmitNoExecute`
+* Potential values:
+ * `strict`
+ * `preferred`
+
+If `strict`, all pods that match this taint must request `memory` that fits it's assigned
+NUMA node `memory` allocatable.
+
+If `preferred`, all pods that match this taint are not required to have their `memory` request
+fit it's assigned NUMA node `memory` allocatable.
+
+## Pod Specification
+
+### API changes
+
+The following API changes are proposed to the `PodSpec` to allow CPU and NUMA affinity to be defined.
+
+```
+// PodSpec is a description of a pod
+type PodSpec struct {
+...
+	// NodeName is a request to schedule this pod onto a specific node.  If it is non-empty,
+	// the scheduler simply schedules this pod onto that node, assuming that it fits resource
+	// requirements.
+	// +optional
+	NodeName string
+	// Identifies a NUMA node that affines the pod.  If it is non-empty, the value must
+	// correspond to a particular NUMA node on the same node that the pod is scheduled against.
+	// This value is only set if either the `CPUAffinity` or `NUMACPUAffinity` tolerations
+	// are present on the pod.
+	// +optional
+	NUMANodeID string
+	// CPUAffinity controls the CPU affinity of the executed pod.
+	// If it is non-empty, the value must correspond to a particular set
+	// of CPU cores in the matching NUMA node on the machine that the pod is scheduled against.
+	// This value is only set if either the `CPUAffinity` or `NUMACPUAffinity` tolerations
+	// are present on the pod.
+	// The values are expressed in the List Format syntax specified here:
+	// here: http://man7.org/linux/man-pages/man7/cpuset.7.html
+	// +optional
+	CPUAffinity string	
+```
+
+### REST API changes
+
+The `/pod/<pod-name>/bind` operation will allow updating the NUMA and CPU
+affinity values.  The same permissions required to schedule a pod to a
+node in the cluster will be required to bind a pod to a particular NUMA node
+and CPU set.
+
+### Tolerations
+
+Pods that require CPU and NUMA affinity prior to execution must set the
+appropriate `Tolerations` for the associated taints.
+
+### Multiple containers
+
+If a pod has multiple containers, the set of containers must all fit
+a specific NUMA node, and the set of affined CPUs are shared among containers.
+
+Pod level cgroups are used to actually affine the container to the specified
+CPU set.
```

Comment:
Is there a proposal or other reference on how pod-level cgroups will work and programs can understand their assignments?

We've had a long-standing discussion around how can a C++ program discover/understand it's resource boundaries so that it can operate properly within them.  A classic third-party example is a Java JVM auto-discovering how many GC threads it can create and how much memory it's been allocated.  It would be desirable to have a path for codebases that may have previously assumed they had to manage a whole machine's worth of resources and handle setting their own scheduler affinities today to be able to understand their partial-machine resource set so they can make their own proper suballocations.